### PR TITLE
Downgrade System.Diagnostics.DiagnosticSource to v6 and add a check if OpenTelemetry instrumentation can be used

### LINKED ISSUE #4456: [Bug] Dependency on System.Diagnostics.DiagnosticSource (>= 7.0.2)  in version 4.58.0 missing in .NET 6 Azure Functions
### Discussed in https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/discussions/4455

<div type='discussions-op-text'>

<sup>Originally posted by **Ozindar** December  6, 2023</sup>
We've got a .NET 6.0 Azure Function using the Identity.Client. This worked fine with 4.57. With a recently deploy we've upgraded to 4.58 but in Azure this triggers a File not found when it's searching for the DiagnosticsSource version 7 assembly.  
"dotnet publish" does not seem put that dll in the output folder. For the moment I've pinned the Identity.Client to 4.58.  
Is there a viable solution to keep using the latest Identity.Client in combination with a .NET6 Azure Function and what would be the best approach?</div>

### PR DESCRIPTION
**Pull request by pmaytak**
Fixes #4456. An alternative to https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/4486 solution.

**Changes proposed in this request**
- Downgrade System.Diagnostics.DiagnosticSource to v6, which is the same version as what Azure Functions runtime uses.
- Catching error when instantiating Open Telemetry fails (means it's not available in that environment) and using null telemetry implementation instead.
- See issue for more details.

**Testing**
- Unit tests pass.
- Tested with a dev app.
- Also tested that Activity code from https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/4243/commits/5dc10b22d580254e86c683fc05396c91eb1e9010 works.

### CODE REVIEW

**Comment by gladjohn** [comment]
thanks for referencing this [issue](https://github.com/Azure/azure-functions-host/issues/8938), so even downgrading still causes some issues?

**Comment by pmaytak** [comment]
> thanks for referencing this [issue](https://github.com/Azure/azure-functions-host/issues/8938), so even downgrading still causes some issues?

Currently downgrading doesn't cause issues. However, it's not future-proof if we ever need newer APIs. Additionally, if the client app enables exporters, OpenTelemetry libraries 1.4.0 and above depend on DiagnosticSource 7.0.0+, so they'll have to use one of the workarounds anyway.

**Review by bgavrilMS — APPROVED** [review]

File: Directory.Packages.props [hunk 1]
```diff
@@ -15,7 +15,7 @@
     <PackageVersion Include="Microsoft.IdentityModel.Abstractions" Version="6.22.0" />
     <PackageVersion Include="Microsoft.Web.WebView2" Version="1.0.864.35" />
     <PackageVersion Include="System.ComponentModel.TypeConverter" Version="4.3.0" />
-    <PackageVersion Include="System.Diagnostics.DiagnosticSource" Version="7.0.2" />
+    <PackageVersion Include="System.Diagnostics.DiagnosticSource" Version="6.0.1" />
```
Comment (bgavrilMS) [inline Directory.Packages.props:18]: Maybe add a comment here about Az Functions issue.

**Comment by bgavrilMS** [comment]
Have you thought about merging both PRs? So that it fixes the assembly load and fallbacks when Az Functions moves from 6 to 8 (?). Alternatively, this PR is enough and when Az Functions updates, the upgrade will break some folks. This should be a strong signal that they need to figure out a better solution. Thoughts?

**Comment by neha-bhargava** [comment]
Were you able to test this with the exporter manually?

**Comment by pmaytak** [comment]
> Have you thought about merging both PRs? So that it fixes the assembly load and fallbacks when Az Functions moves from 6 to 8 (?). Alternatively, this PR is enough and when Az Functions updates, the upgrade will break some folks. This should be a strong signal that they need to figure out a better solution. Thoughts?

Yea, I'll merge the PRs.

You mean when Azure Functions update their runtime to v8? I tested with a v5 library and everything works. So if they upgrade to v9 and we use v6/7, it's okay.

> Were you able to test this with the exporter manually?

Yes, with the dev app. Workarounds still needed for in-process Function, but works otherwise.

**Review by gladjohn — APPROVED** [review]

**Review by neha-bhargava — APPROVED** [review]
I did a quick test with MISE exporter too and logs look fine. Thanks!

**Pull request merged**